### PR TITLE
Allow installation with php 8

### PR DESCRIPTION
**Pull request by alexander-schranz**
## Summary of the changes / Why this is an improvement

Requires I think release of Crate-PDO before this can be merged: https://github.com/crate/crate-pdo/issues/115

## Checklist

 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

**Comment by alexander-schranz** [comment]
Currently the dbal seems not to be compatible because of breaking change on there side: https://github.com/doctrine/dbal/issues/4619

**Comment by amotl** [comment]
CI/GHA responds with:

> Warning: The lock file is not up to date with the latest changes in `composer.json`. You may be getting outdated dependencies. It is recommended that you run `composer update`.

Can you invoke `composer update --lock` with the minimum designated PHP version? So, while the repository is still on PHP7.2, you can well move on to PHP7.3, because `crate-pdo` did also drop support for PHP7.2 when moving on to PHP8.

**Comment by amotl** [comment]
[1] croaks with dependency errors. You will need to use `"crate/crate-pdo": "^2.0.0"` in `composer.json`. Don't forget to lock!

[1] https://github.com/crate/crate-dbal/runs/2452547572#step:7:7

**Comment by amotl** [comment]
> Your lock file does not contain a compatible set of packages. Please run composer update.
> 
> Problem 1
> - crate/crate-pdo is locked to version 1.1.0 and an update of this package was not requested.
> - crate/crate-pdo 1.1.0 requires php ^7.2 -> your php version (8.0.3) does not satisfy that requirement.

Probably need to drop PHP 7.2 and depend on `crate-pdo 2.0.0` only.

[2] https://github.com/crate/crate-dbal/runs/2452605795#step:7:6

**Comment by alexander-schranz** [comment]
@amotl normally libraries in PHP do not commit `.lock` files and the composer install the depending on the `composer.json` the correct dependencies. The lock file is only used by projects and a library composer.lock file does not have any effect on a project using it so it can safely be removed.

**Comment by amotl** [comment]
> Normally _libraries_ in PHP **do not** commit `.lock` files and the composer install the depending on the `composer.json` the correct dependencies. The lock file is only used by _projects_ and a library `composer.lock` file does not have any effect on a project using it so it can safely be removed.

All right, I now see that also at [1,2]. I recently argued the other way round (see #115), but I admit I was apparently wrong and didn't make the distinction between drivers and projects appropriately. Mea culpa.

So, let's remove the `.lock` files right away on both projects? Please note that `.github/workflows/tests.yml` will also have to be adjusted to use `composer.json` in line 49.

[1] https://github.com/doctrine/dbal
[2] https://github.com/doctrine/orm

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -36,21 +36,10 @@ jobs:
...
 
-      # Cache Composer Dependencies
-      # https://github.com/marketplace/actions/setup-php-action#cache-composer-dependencies
-      - name: Get composer cache directory
-        id: composer-cache-dir
-        run: echo "::set-output name=dir::$(composer config cache-dir)"
-      - name: Apply dependency caching
-        id: composer-cache
-        uses: actions/cache@v2
+      - uses: "ramsey/composer-install@v1"
```
Comment (alexander-schranz) [inline .github/workflows/tests.yml:39]: takes care of the caching logic

Comment (amotl) [reply .github/workflows/tests.yml:39]: https://github.com/ramsey/composer-install looks good to me, thank you.

**Comment by amotl** [comment]
Did you know we added support for invoking the test suite easily by using Docker for running CrateDB [1] just the other day?

[1] https://github.com/crate/crate-dbal/blob/main/DEVELOP.rst#using-docker

**Comment by alexander-schranz** [comment]
Think https://github.com/crate/crate-dbal/pull/113 is required before we can continue here.

File: composer.json [hunk 2]
```diff
@@ -34,6 +34,11 @@
             "Doctrine\\Tests": "vendor/doctrine/dbal/tests"
         }
     },
+    "config": {
+        "preferred-install": {
+            "doctrine/dbal": "source"
```
Comment (alexander-schranz) [inline composer.json:39]: required to test crate

File: composer.json [hunk 3]
```diff
@@ -34,6 +34,11 @@
             "Doctrine\\Tests": "vendor/doctrine/dbal/tests"
         }
     },
+    "config": {
```
Comment (alexander-schranz) [inline composer.json:37]: the config attribute is root-only so does not has any effect on projects using this library https://getcomposer.org/doc/04-schema.md#config

**Comment by amotl** [comment]
https://github.com/crate/crate-dbal/runs/2453728629#step:7:5 shows the same as observed at https://github.com/crate/crate-dbal/pull/116#issuecomment-828085703. Maybe something related to the new `ramsey/composer-install` action on GHA?

```
PHP Fatal error:  Uncaught Error: Class 'Doctrine\Tests\DBAL\Platforms\AbstractPlatformTestCase' not found in /home/runner/work/crate-dbal/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php:40
Stack trace:
#0 /home/runner/work/crate-dbal/crate-dbal/vendor/phpunit/phpunit/src/Util/FileLoader.php(66): include_once()
#1 /home/runner/work/crate-dbal/crate-dbal/vendor/phpunit/phpunit/src/Util/FileLoader.php(54): PHPUnit\Util\FileLoader::load()
#2 /home/runner/work/crate-dbal/crate-dbal/vendor/phpunit/phpunit/src/Framework/TestSuite.php(384): PHPUnit\Util\FileLoader::checkAndLoad()
#3 /home/runner/work/crate-dbal/crate-dbal/vendor/phpunit/phpunit/src/Framework/TestSuite.php(482): PHPUnit\Framework\TestSuite->addTestFile()
#4 /home/runner/work/crate-dbal/crate-dbal/vendor/phpunit/phpunit/src/Util/Configuration.php(1031): PHPUnit\Framework\TestSuite->addTestFiles()
#5 /home/runner/work/crate-dbal/crate-dbal/vendor/phpunit/phpunit/src/Util/Configuration.php(900): PHPUnit\Util\Configuration->getTestSuite()
#6 /home/runner/w in /home/runner/work/crate-dbal/crate-dbal/test/Crate/Test/DBAL/Platforms/CratePlatformTest.php on line 40
```

File: composer.json [hunk 4]
```diff
@@ -15,7 +15,7 @@
     "minimum-stability": "dev",
     "prefer-stable": true,
     "require": {
-        "php": "^7.3",
+        "php": "^7.3 || ^8.0",
         "doctrine/dbal": "^2.10",
```
Comment (amotl) [inline composer.json:19]: Maybe this needs 2.13 now?

**Comment by alexander-schranz** [comment]
Seems like I need to change https://github.com/crate/crate-pdo/pull/125 to match the return types of doctrine dbal.

And need then use the doctrine dbal PDOStatementImplementations

**Comment by amotl** [comment]
This just happened again at https://github.com/crate/crate-dbal/runs/2453953436:
```
PHP Fatal error:  Uncaught Error: Class 'Doctrine\Tests\DBAL\Platforms\AbstractPlatformTestCase' not found
```

**Comment by amotl** [comment]
Tests pass, sweet. Can you rebase?

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -50,13 +52,13 @@ jobs:
       # https://github.com/codecov/codecov-action
       - name: Upload coverage results to Codecov
         uses: codecov/codecov-action@v1
-        if: always() && (matrix.php-version == '7.4' || matrix.php-version == '8.0')
+        if: always() && (matrix.php-version == '7.4')
```
Comment (amotl) [inline .github/workflows/tests.yml:55]: Can you bring this back and also adjust `.scrutinizer.yml` to say `runs: 2` again?

**Review by amotl — CHANGES_REQUESTED** [review]

**Comment by alexander-schranz** [comment]
@amotl not sure why the prefer-source is ignored when loading something from cache directory. I now forced prefer-source by installing doctrine/dbal manually again over the source mode. Hope this fixes it for a longer time.

Branch is rebased!

File: .github/workflows/tests.yml (hunk 5 above)
Comment (alexander-schranz) [reply .github/workflows/tests.yml]: code coverage is back needed to use phpunit 9 for it as 8 didn't support code coverage for php 8

**Comment by amotl** [comment]
Thanks a stack!

**Comment by alexander-schranz** [comment]
@amotl Thank you for the support!

**Comment by amotl** [comment]
https://packagist.org/packages/crate/crate-dbal#3.0.0 has been released. Thank you again for making this happen and let me know about any flaws you might observe.

**Pull request merged**